0.9.2 panic instead of cycle error

I just upgraded my terraform plans to 0.9.2 from 0.8.2. In 0.8.2 I encountered a cycle error when running a destroy, however, in 0.9.2 the cycle error is replaced with a panic.

### Terraform Version
Terraform v0.9.2

### Panic Output
Crash.log: https://gist.github.com/gerr1t/eea1fdc318f711dff17f3f98cd2edcfd

```
Releasing state lock. This may take a few moments...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x159f196]

goroutine 1957 [running]:
github.com/hashicorp/terraform/terraform.(*State).sort(0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/state.go:745 +0x26
github.com/hashicorp/terraform/terraform.WriteState(0x0, 0x8050700, 0xc4201052d0, 0x1038a15, 0xc420cbbd10)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/state.go:2000 +0x40
github.com/hashicorp/terraform/state/remote.(*State).PersistState(0xc420750060, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/state/remote/state.go:58 +0x79
github.com/hashicorp/terraform/backend/local.(*Local).opApply(0xc4201e48c0, 0x8075f80, 0xc42076a7c0, 0xc4201fc790, 0xc420831590)
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend_apply.go:138 +0x5d9
github.com/hashicorp/terraform/backend/local.(*Local).(github.com/hashicorp/terraform/backend/local.opApply)-fm(0x8075f80, 0xc42076a7c0, 0xc4201fc790, 0xc420831590)
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend.go:226 +0x52
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc4201e48c0, 0xc42064cb00, 0xc42064caf0, 0x8075f80, 0xc42076a7c0, 0xc4201fc790, 0xc420831590)
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend.go:246 +0x9a
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend.go:247 +0x192
```

### Expected Behavior
Either the cycle error to be gone, or to present me with a cycle error.

### Actual Behavior
A panic.

--- a/backend/local/backend_apply.go
+++ b/backend/local/backend_apply.go
@@ -102,7 +102,9 @@ func (b *Local) opApply(
 	doneCh := make(chan struct{})
 	go func() {
 		defer close(doneCh)
-		applyState, applyErr = tfCtx.Apply()
+		_, applyErr = tfCtx.Apply()
+		// we always want the state, even if apply failed
+		applyState = tfCtx.State()
 
 		/*
 			// Record any shadow errors for later

--- a/terraform/context.go
+++ b/terraform/context.go
@@ -453,8 +453,17 @@ func (c *Context) Input(mode InputMode) error {
 // Apply applies the changes represented by this context and returns
 // the resulting state.
 //
-// In addition to returning the resulting state, this context is updated
-// with the latest state.
+// Even in the case an error is returned, the state may be returned and will
+// potentially be partially updated.  In addition to returning the resulting
+// state, this context is updated with the latest state.
+//
+// If the state is required after an error, the caller should call
+// Context.State, rather than rely on the return value.
+//
+// TODO: Apply and Refresh should either always return a state, or rely on the
+//       State() method. Currently the helper/resource testing framework relies
+//       on the absence of a returned state to determine if Destroy can be
+//       called, so that will need to be refactored before this can be changed.
 func (c *Context) Apply() (*State, error) {
 	defer c.acquireRun("apply")()
 
@@ -580,7 +589,7 @@ func (c *Context) Plan() (*Plan, error) {
 // to their latest state. This will update the state that this context
 // works with, along with returning it.
 //
-// Even in the case an error is returned, the state will be returned and
+// Even in the case an error is returned, the state may be returned and
 // will potentially be partially updated.
 func (c *Context) Refresh() (*State, error) {
 	defer c.acquireRun("refresh")()

--- a/terraform/state.go
+++ b/terraform/state.go
@@ -1960,11 +1960,11 @@ func ReadStateV2(jsonBytes []byte) (*State, error) {
 		}
 	}
 
-	// Sort it
-	state.sort()
-
 	// catch any unitialized fields in the state
 	state.init()
+
+	// Sort it
+	state.sort()
 
 	return state, nil
 }
@@ -1995,11 +1995,11 @@ func ReadStateV3(jsonBytes []byte) (*State, error) {
 		}
 	}
 
-	// Sort it
-	state.sort()
-
 	// catch any unitialized fields in the state
 	state.init()
+
+	// Sort it
+	state.sort()
 
 	// Now we write the state back out to detect any changes in normaliztion.
 	// If our state is now written out differently, bump the serial number to
@@ -2020,11 +2020,16 @@ func ReadStateV3(jsonBytes []byte) (*State, error) {
 
 // WriteState writes a state somewhere in a binary format.
 func WriteState(d *State, dst io.Writer) error {
-	// Make sure it is sorted
-	d.sort()
+	// writing a nil state is a noop.
+	if d == nil {
+		return nil
+	}
 
 	// make sure we have no uninitialized fields
 	d.init()
+
+	// Make sure it is sorted
+	d.sort()
 
 	// Ensure the version is set
 	d.Version = StateVersion

